### PR TITLE
feat: show info on missing metadata

### PR DESCRIPTION
**Pull request by johanneskoester**
In dryrun, if there is no metadata recorded for certain jobs, we now print an info.

### QC

* [x] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [x] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of Snakemake).

## Summary by CodeRabbit

- **New Features**
	- Introduced a function to extract and sort rule names from job instances, enhancing reporting capabilities.
	- Added a method to verify the presence of metadata for job outputs, improving job handling robustness.
	- Implemented logging for jobs that lack recorded metadata or provenance, providing better visibility.

- **Bug Fixes**
	- Enhanced control flow in job metadata evaluation to prevent unnecessary checks when metadata is absent.

- **Chores**
	- Improved type safety and clarity through the addition of type hinting.

### CODE REVIEW

File: snakemake/workflow.py [hunk 1]
```diff
@@ -1143,6 +1144,18 @@ def execute(
...
+                no_metadata_jobs = [
+                    job for job in self.dag.jobs if self.dag.reason(job).no_metadata
+                ]
+                if no_metadata_jobs:
+                    logger.info(
+                        "Some jobs have no recorded provenance/metadata so that they "
+                        "cannot be triggered by this information. \n"
+                        "Rules with missing "
+                        f"metadata: {' '.join(jobs_to_rulenames(no_metadata_jobs))}"
+                    )
```
[the remark this answers is not in the document]
Comment (johanneskoester) [reply snakemake/workflow.py]: The trailing comma is not needed, we use black for formatting and it does not want it here.
The type annotation is also unnecessary as this is an internal helper function.

File: snakemake/dag.py [hunk 2]
```diff
@@ -1257,30 +1257,36 @@
...
+                                        self.workflow.persistence.code_changed(job)
+                                    )
+                                if (
+                                    RerunTrigger.SOFTWARE_ENV
+                                    in self.workflow.rerun_triggers
+                                ):
+                                    reason.software_stack_changed = any(
+                                        self.workflow.persistence.conda_env_changed(job)
+                                    ) or any(
+                                        self.workflow.persistence.container_changed(job)
```
[the remark this answers is not in the document]
Comment (johanneskoester) [reply snakemake/dag.py:1288]: No, we use black for formatting and it does not want it here.

File: snakemake/workflow.py [hunk 3]
```diff
@@ -1230,11 +1246,7 @@ def log_provenance_info():
                     )
                     logger.info(
                         "Rules with provenance triggered jobs: "
-                        + ",".join(
-                            sorted(
-                                set(job.rule.name for job in provenance_triggered_jobs)
-                            )
-                        )
+                        + " ".join(jobs_to_rulenames(provenance_triggered_jobs))
                     )
```
[the remark this answers is not in the document]
Comment (johanneskoester) [reply snakemake/workflow.py:1251]: Never hint on trailing commas.

File: snakemake/workflow.py [hunk 4]
```diff
@@ -1143,6 +1144,20 @@
                 # no shared FS, hence we have to upload the sources to the storage
                 self.upload_sources()
 
+            def log_missing_metadata_info():
```
[the remark this answers is not in the document]
Comment (johanneskoester) [reply snakemake/workflow.py:1147]: Do not hint on type annotations for functions that are inside of methods.

File: snakemake/dag.py [hunk 5]
```diff
@@ -1257,30 +1257,39 @@ async def update_needrun(job):
...
-                            ):
-                                reason.software_stack_changed = any(
-                                    self.workflow.persistence.conda_env_changed(job)
-                                ) or any(
-                                    self.workflow.persistence.container_changed(job)
-                                )
+
+                            if not self.workflow.persistence.has_metadata(job):
+                                reason.no_metadata = True
+                            else:
```
Comment (laf070810) [inline snakemake/dag.py:1263]: This change causes a bug: when using `--drop-metadata` (for filesystem speed), changes except file mtime (especially mtime of `script`/`notebook`) no longer trigger the workflow. I think we should handle this case, or make `no_metadata` a choice of RerunTrigger.

Comment (laf070810) [reply snakemake/dag.py:1263]: Proposed a fix in #3148

**Pull request merged**